### PR TITLE
(4.x.x) Add DEFLATE compression (RFC 1950 / 1951) to compression module

### PR DESCRIPTION
**Pull request by chakl**
------
### Description:

This adds DEFLATE compression (RFC 1950 / 1951) to the existdb compression module.

### Reference:

RFC 1951 DEFLATE is required for the existdb SAMLv2 implementation in XQuery.
Ref: Section 3.4.4.1 "DEFLATE Encoding" in SAMLv2 standard documentation https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf

While there, implement RFC 1950 as well.

### Type of tests:

I'm not familiar with the testing framework. This PR includes an example XQuery "example-deflate.xql" that demonstrates usage and tests compression/uncompression against its own behavior. I can add interoperability tests against Perl's DEFLATE implementation, not sure how.

### Misc

License follows the existing existdb license.

### CODE REVIEW

File: extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	boolean rawflag = false;
+        if(args.length > 1 && !args[1].isEmpty())
+	    rawflag = args[1].itemAt(0).convertTo(Type.BOOLEAN).effectiveBooleanValue();
+
+	Deflater defl = new Deflater(java.util.zip.Deflater.DEFAULT_COMPRESSION, rawflag);
+
+        // deflate the data
+        try(final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+	    DeflaterOutputStream dos = new DeflaterOutputStream(baos, defl)) {
+	    //DeflaterOutputStream dos = new DeflaterOutputStream(baos, new Deflater(java.util.zip.Deflater.DEFAULT_COMPRESSION, true))) {
```
Comment (dizzzz) [inline extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: please remove the comment line

Comment (chakl) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: oh, missed that line. will do

File: extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        // deflate the data
+        try(final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+	    DeflaterOutputStream dos = new DeflaterOutputStream(baos, defl)) {
+            bin.streamBinaryTo(dos);
+            
+            dos.flush();
+            dos.finish();
+            
+            return BinaryValueFromInputStream.getInstance(context, new Base64BinaryValueType(), new ByteArrayInputStream(baos.toByteArray()));
```
[the remark this answers is not in the document]
Comment (chakl) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: thanks, need to check docs

[the remark this answers is not in the document]
Comment (adamretter) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: You should consider the `FastByteArrayOutputStream` and `FastByteArrayInputStream` classes.

File: extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this library; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * $Id$
+ */
+package org.exist.xquery.modules.compression;
+
+import java.io.ByteArrayInputStream;
```
Comment (adamretter) [inline extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: Could you use the eXist-db variant `org.exist.util.io.FastByteArrayInputStream` please?

File: extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import java.io.IOException;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterOutputStream;
+
+import org.apache.commons.io.output.ByteArrayOutputStream;
```
Comment (adamretter) [inline extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: Could you use the eXist-db variant `org.exist.util.io.FastByteArrayOuputStream` please?

**Comment by adamretter** [comment]
@chakl When it's ready, do you know how to send PR's for the ports? If not, I can help explain how to do that...

File: extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chakl) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: will check, saw that in your code

File: extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java (hunk 3 above)
Comment (dizzzz) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: ow yea, sorry I forgot about that one..... does it have an efficient `toInputStream()` too? (meaning: no serialization to another `byte[]`)

Comment (adamretter) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: @dizzzz `FastByteArrayInputStream` is an `InputStream`, so why would you call `toInputStream`?

Comment (dizzzz) [reply extensions/modules/src/org/exist/xquery/modules/compression/DeflateFunction.java]: ahhhh sorry, coffee time.....

**Comment by chakl** [comment]
@adamretter Not sure what you mean "send PR's for the ports". I have changed the code to use FastByteArrayOutputStream as you suggested, tested again and committed.

**Comment by adamretter** [comment]
@chakl thanks again for your contribution :-)

**Pull request merged**